services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureSessionEnforceExecutor.java

```diff
@@ -0,0 +1,87 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.services.clientpolicy.executor;
+
+import org.jboss.logging.Logger;
+import org.keycloak.OAuthErrorException;
+import org.keycloak.component.ComponentModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.protocol.oidc.endpoints.request.AuthorizationEndpointRequest;
+import org.keycloak.protocol.oidc.utils.OIDCResponseType;
+import org.keycloak.services.clientpolicy.AuthorizationRequestContext;
+import org.keycloak.services.clientpolicy.ClientPolicyContext;
+import org.keycloak.services.clientpolicy.ClientPolicyException;
+import org.keycloak.services.clientpolicy.ClientPolicyLogger;
+import org.keycloak.util.TokenUtil;
+
+public class SecureSessionEnforceExecutor implements ClientPolicyExecutorProvider {
+
+    private static final Logger logger = Logger.getLogger(SecureSessionEnforceExecutor.class);
+
+    private final KeycloakSession session;
+    private final ComponentModel componentModel;
+
+    public SecureSessionEnforceExecutor(KeycloakSession session, ComponentModel componentModel) {
+        this.session = session;
+        this.componentModel = componentModel;
+    }
+
+    @Override
+    public String getName() {
+        return componentModel.getName();
+    }
+
+    @Override
+    public String getProviderId() {
+        return componentModel.getProviderId();
+    }
+
+    @Override
+    public void executeOnEvent(ClientPolicyContext context) throws ClientPolicyException {
+        switch (context.getEvent()) {
+            case AUTHORIZATION_REQUEST:
+                AuthorizationRequestContext authorizationRequestContext = (AuthorizationRequestContext)context;
+                executeOnAuthorizationRequest(authorizationRequestContext.getparsedResponseType(),
+                    authorizationRequestContext.getAuthorizationEndpointRequest(),
+                    authorizationRequestContext.getRedirectUri());
+                return;
+            default:
+                return;
+        }
+    }
+
+    private void executeOnAuthorizationRequest(
+            OIDCResponseType parsedResponseType,
+            AuthorizationEndpointRequest request,
+            String redirectUri) throws ClientPolicyException {
+        ClientPolicyLogger.log(logger, "Authz Endpoint - authz request");
+        if (TokenUtil.isOIDCRequest(request.getScope())) {
+            if(request.getNonce() == null) {
+                ClientPolicyLogger.log(logger, "Missing parameter: nonce");
+                throw new ClientPolicyException(OAuthErrorException.INVALID_REQUEST, "Missing parameter: nonce");
+            }
+        } else {
+            if(request.getState() == null) {
```

Comment:
Just to doublechec, wouldn't it be better to check that "state" parameter is always there? In other words, remove it from the "else" branch? This would mean that for OIDC clients, both "state" and "nonce" will be enforced, which IMO is fine behaviour?

Reply:
@mposolda Yes, as you pointed out, it is more secure for OIDC clients to be enforced with both "state" and "nonce".
However, considering passing all FAPI-RW conformance tests, it should not be to do that. The one test case of current FAPI-RW conformance tests "fapi-rw-id2-ensure-authorization-request-without-state-success" checks whether the authorization server accepts the authorization request without "state" from OIDC client.

Reply:
Ok, Thanks for explaining that.